### DISCUSSION TITLE
Allow Access to Protected Class Members

### DISCUSSION THREAD

**Discussion by damepatr** [moq/labs#98] [Ideas]
Hi folks! 👋

# Summary
As far as I know, _Moq_ doesn't provide direct access to `protected` class members. You can call `ProtectedExtension.Protected<T>(Mock<T> mock)` to setup the mock's `protected` members, but you can't directly get/set `protected` properties/fields, and you can't directly call `protected` methods. This can make it difficult to mock a base class and test the base class's functionality.

# Example
Suppose that you have the following:
```
public abstract class BaseClass {
    protected string _commonField;

    protected BaseClass() {
        _commonField = GetStringWithId("SomeId");
    }

    public string GetStringWithId(string id) {
        // Returns string with given ID.
    }
}

public class FirstImplementation : BaseClass  {
    public FirstImplementation() : base() { }

    public void SomeMethod() {
        // Uses _commonField.
    }
}

public class SecondImplementation : BaseClass {
    public SecondImplementation() : base() { }

    public void AnotherMethod() {
        // Uses _commonField
    }
}
```

Since `_commonField` is used by implementors, I would argue that you should create a unit test for the `_commonField` property. Ideally, you would do this by using the partial mock feature with `BaseClass`. This unit test would validate that `_commonField` is getting set correctly in the constructor; that way you can validate that the correct value is getting provided to `FirstImplementation` and `SecondImplementation`.

The concern is that _Moq_ doesn't provide direct access to the `_commonField` field, so you can't verify that it is getting set correctly. Instead, you either have to (a) create your own mock or (b) use reflection.

**Comment by kzu** [comment]
This is not a Moq issue, but a design issue. A protected field (non-readonly even!) can be overwritten by anyone at any point. Not a great design to preserve the invariants of the class.

If you need to expose internal state just for verification in tests, I'd suggest creating internal members that expose whatever you need and perform regular state assertions on the resulting (mock) object itself. You can make those members visible with InternalsVisibleTo easily.

That said, having your tests so deeply tied to internal object state is a recipe for brittle tests that need to be constantly changed whenever implementations (inevitably) evolve.

**Reply by damepatr** [reply]
@kzu 

> A protected field (non-readonly even!) can be overwritten by anyone at any point.

But sometimes that is what you want to do. Example: You may be writing a library that publishes an abstract class, and implementors can inherit and override the protected methods of that abstract class to get their desired functionality.

**Reply by damepatr** [reply]
> If you need to expose internal state just for verification in tests, I'd suggest creating internal members that expose whatever you need and perform regular state assertions on the resulting (mock) object itself. You can make those members visible with InternalsVisibleTo easily.

Users of Moq may not want to expose members to other files in the same assembly. This wouldn't be an ideal solution for them.

**Reply by damepatr** [reply]
Apologies if this comes across as a bit contentious. I'll try to come up with another example that may make a little more sense.

**Reply by kzu** [reply]
If you're designing a library for third-parties to consume and are exposing protected fields, you're opening yourself to private state corruption since anyone can set those fields, violating invariants that are now entirely out of your control. That's just bad OOP design. I'd suggest reading a bit on the concept of encapsulation.

**Reply by damepatr** [reply]
Imagine that you have an abstract class like this:
``` C#
public abstract class Item
    {
        protected enum WeightClass
        {
            Heavy,
            Light
        }

        public int Weight { get; }

        protected WeightClass _weightClass { get => Weight >= 50 ? WeightClass.Heavy : WeightClass.Light; }

        protected Item(int weight)
        {
            if (weight < 0) throw new ArgumentOutOfRangeException(nameof(weight));
            this.Weight = weight;
        }
    }
```

**Reply by damepatr** [reply]
You can inherit from this class to create a specific item as shown below:
``` C#
public class Pickaxe : Item
{
	public Pickaxe() : base(10)
	{
	}

	public void EquipTo(Entity entity)
	{
		// Uses _equipTime.
	}

	private TimeSpan _equipTime
	{
		get
		{
			// Uses _weightClass.
		}
	}
}
```

**Reply by damepatr** [reply]
`Weight` and `_weightClass` in `Item` can't be overridden because they are non-virtual properties. Each implementation also must provide a valid value for `weight`.

This means that `_weightClass` is always going to be a valid value. You can use it in implementations of `Item` without concern.

**Reply by damepatr** [reply]
But now let's say that you want to test the `_weightClass` property since it is used by implementations. `Moq` doesn't provide access to that property, so you would have to create your own mock. That isn't a very big deal in this example, but it can become tedious if `Item` has many abstract methods.

**Reply by kzu** [reply]
Now you made it a property. So now it works OOB with Moq 😉 . See https://github.com/Moq/moq4/wiki/Quickstart#miscellaneous (there's a way to even define an interface for protected members to get nicer intellisense).

Naming conventions in .NET would frown at a protected *property* with an underscore prefix and non-PascalCase notation, btw.

**Reply by damepatr** [reply]
@kzu - You can setup the callbacks and returns of the `_weightClass` property, but how do you get the property in a unit test? Let's say that you are writing the following unit test:
``` C#
[TestMethod]
public void WeightClass_WeightLessThan50_ReturnsLight()
{
	var mock = new Mock<Item>(40) { CallBase = true };

        // How do we get the _weightClass property from the mock?
	var weightClass = -1;

	Assert.AreEqual(WeightClass.Light, weightClass);
}
```

How do we get the protected `_weightClass` property from the mock? Do you have to use reflection?

**Reply by kzu** [reply]
That's a general C#/.NET question, not specific to Moq.

You'd make internal protected so tests can read it directly (by having an InternalsVisibleTo to the test project.

**Reply by damepatr** [reply]
Ok, I think that I see now. You can use reflection, `internal protected`, or a manual mock to expose the protected `_weightClass` property.

**Reply by damepatr** [reply]
My initial thought process was that we could add a way to get the protected `_weightClass` property. An advantage of manual mocking is that you can add public members which expose protected members. The issue is that `Moq.Mock<T>.Object` returns type `T`, and that means that we can't add public members.

**Reply by damepatr** [reply]
You could potentially return an object that inherits from `T` but provides additional members for accessing protected members. This could be done with the [`TypeBuilder`](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.typebuilder?view=net-7.0) class, but it seems complicated. It would probably be best to keep things simple.

**Reply by kzu** [reply]
You just make a Mock<T> of the derived type that exposes the additional protected members as public members and just passes whatever ctor args to the base ctor. Just like you would with a manual mock too, btw.

**Reply by damepatr** [reply]
> You just make a Mock of the derived type that exposes the additional protected members as public members and just passes whatever ctor args to the base ctor. Just like you would with a manual mock too, btw.

Yeah, this makes sense. I think that this is preferable to exposing `protected` members through the [TypeBuilder](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.typebuilder?view=net-7.0) class.
